src/script_opt/FuncInfo.cc

```diff
@@ -6,461 +6,507 @@
 
 namespace zeek::detail {
 
-// The following BiFs do not have any script-level side effects. It's
-// followed by comments listing the BiFs that have been omitted, and why.
-//
-// See script_opt/ZAM/maint/README for maintenance of these lists.
+// See script_opt/ZAM/maint/README for maintenance of the attributes
+// in this file.
+
+// Attributes associated with functions. Currently these are mutually
+// exclusive (i.e., no function will have more than one), but for now
+// we use a bitmask-style approach so we can accommodate future attributes
+// that might overlap.
+
+// BiF Functions that are not listed are assumed to have Unknown side effects.
+// (These are described in comments after the table definition.)  Script
+// functions that are not listed as assumed to not be "special", i.e. known
+// to the event engine.
+
+// Does not change script-level state (though may change internal state).
+#define ATTR_NO_SCRIPT_SIDE_EFFECTS 0x1
```

Comment:
Nothing to address and it's certainly subjective, but positive terms are generally easier to reason about and help avoiding double negatives.

```
if ( ! has_no_script_side_effects(b->Name()) )
            return false;
```
For someone just skimming this and not immersed, it's mind-twisting :-)

Reply:
likewise, noted this for next month's maintenance